File: investor/projects_engine.py
"""Investment project evaluation.

Given a project (list of cost items + funding sources) and a scenario,
compute whether the chosen funding (at the project's target date in that
scenario) is sufficient to cover the cost — accounting for tax on the
liquidations.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import date
from typing import Any, Optional

from .fx import convert
from .growth import GrowthRateService
from .models import (
    FundingSource,
    InvestmentProject,
    Property,
    Scenario,
    Settings,
    StockHolding,
)
from .projections import project_property_value_at, project_stock_value_at
from .tax import property_sale_tax, stock_sale_tax, years_between


@dataclass
class FundingLineResult:
    kind: str
    asset_label: str
    gross_proceeds: float
    tax: float
    net_proceeds: float
    detail: dict[str, Any] = field(default_factory=dict)


@dataclass
class ProjectEvaluation:
    project_id: str
    project_name: str
    scenario_id: Optional[str]
    scenario_name: Optional[str]
    target_date: Optional[str]
    primary_currency: str
    total_cost: float
    funding_lines: list[FundingLineResult] = field(default_factory=list)
    total_gross: float = 0.0
    total_tax: float = 0.0
    total_net_funding: float = 0.0
    surplus_or_shortfall: float = 0.0  # positive = surplus, negative = shortfall

    @property
    def is_funded(self) -> bool:
        return self.surplus_or_shortfall >= 0


def _resolve_target_date(project: InvestmentProject) -> date:
    if project.target_date:
        try:
            return date.fromisoformat(project.target_date)
        except ValueError:
            pass
    return date.today()


def evaluate_project(
    *,
    project: InvestmentProject,
    scenario: Scenario,
    holdings: list[StockHolding],
    properties: list[Property],
    settings: Settings,
    growth_service: Optional[GrowthRateService] = None,
) -> ProjectEvaluation:
    target = _resolve_target_date(project)
    primary_ccy = settings.primary_currency

    # Project total cost in primary currency (items may be in other currencies)
    total_cost = 0.0
    for item in project.items:
        total_cost += convert(item.cost, item.currency, primary_ccy, settings)
    # If the project's currency differs from primary, the "label" still uses project.currency
    # but math is done in primary.

    holdings_by_id = {h.id: h for h in holdings}
    properties_by_id = {p.id: p for p in properties}

    lines: list[FundingLineResult] = []

    for fs in project.funding:
        if fs.kind == "cash":
            net = float(fs.amount_or_shares)
            lines.append(
                FundingLineResult(
                    kind="cash",
                    asset_label="Cash",
                    gross_proceeds=net,
                    tax=0.0,
                    net_proceeds=net,
                    detail={"note": "Cash contribution (assumed in primary currency)."},
                )
            )
            continue

        if fs.kind == "stock":
            h = holdings_by_id.get(fs.asset_id or "")
            if not h:
                lines.append(FundingLineResult(kind="stock", asset_label="(missing)", gross_proceeds=0, tax=0, net_proceeds=0,
                                               detail={"error": f"Holding {fs.asset_id} not found"}))
                continue
            shares_to_sell = float(fs.amount_or_shares)

            # Get projected price at target date
            v = project_stock_value_at(h, scenario, target, primary_ccy, settings)
            projected_price_native = v["projected_price"]  # in holding's currency

            # Liquidity check: only liquid shares can actually be sold
            vested = h.vested_shares_at(target)
            second_trigger_iso = scenario.second_trigger_for(h.id, h.second_trigger_date)
            is_double = h.equity_type == "RSU (double trigger)"
            second_trigger_passed = True
            if is_double and second_trigger_iso:
                try:
                    second_trigger_passed = target >= date.fromisoformat(second_trigger_iso)
                except ValueError:
                    pass
            liquid_avail = vested if (not is_double or second_trigger_passed) else 0.0
            shares_actually = min(shares_to_sell, liquid_avail)

            # Cost basis: for RSUs taxed at vest, basis = vest price ≈ current price for already-vested.
            # We approximate using cost_basis_per_share if set; otherwise current price (no gain).
            cost_basis = h.cost_basis_per_share if h.cost_basis_per_share > 0 else h.current_share_price
            # Pragmatic holding period proxy: time from today to target date.
            # For RSUs taxed at vest, basis = vest price, so this measures LT eligibility from vest forward.
            holding_period = max(0.0, (target - date.today()).days / 365.25)

            tx = stock_sale_tax(
                jurisdiction=h.jurisdiction,
                sale_price_per_share=projected_price_native,
                cost_basis_per_share=cost_basis,
                shares=shares_actually,
                holding_period_years=holding_period,
                settings=settings,
            )

            gross_p = convert(tx["gross_proceeds"], h.currency, primary_ccy, settings)
            tax_p = convert(tx["tax"], h.currency, primary_ccy, settings)
            net_p = convert(tx["net_proceeds"], h.currency, primary_ccy, settings)

            label = f"{h.ticker or h.company_name} — {shares_actually:,.0f} sh @ ~{projected_price_native:,.2f} {h.currency}"
            detail = dict(tx)
            detail["projected_price_native"] = projected_price_native
            detail["liquid_shares_available"] = liquid_avail
            detail["shares_requested"] = shares_to_sell
            detail["shares_sold"] = shares_actually
            if shares_actually < shares_to_sell:
                detail["warning"] = (
                    f"Only {liquid_avail:,.0f} shares liquid at {target.isoformat()}; capped from {shares_to_sell:,.0f}."
                )

            lines.append(
                FundingLineResult(
                    kind="stock",
                    asset_label=label,
                    gross_proceeds=gross_p,
                    tax=tax_p,
                    net_proceeds=net_p,
                    detail=detail,
                )
            )
            continue

        if fs.kind == "property":
            p = properties_by_id.get(fs.asset_id or "")
            if not p:
                lines.append(FundingLineResult(kind="property", asset_label="(missing)", gross_proceeds=0, tax=0, net_proceeds=0,
                                               detail={"error": f"Property {fs.asset_id} not found"}))
                continue
            fraction = max(0.0, min(1.0, float(fs.amount_or_shares)))

            # Compute projected value directly in property's native currency
            v_native = project_property_value_at(p, scenario, target, p.currency, settings, growth_service)
            sale_price_native = v_native["gross"] * fraction

            cost_basis_native = p.purchase_price * fraction
            holding_period = years_between(p.purchase_date, target)
            mortgage_payoff_native = p.mortgage_balance * fraction  # crude proportional

            tx = property_sale_tax(
                jurisdiction=p.jurisdiction,
                sale_price=sale_price_native,
                cost_basis=cost_basis_native,
                holding_period_years=holding_period,
                mortgage_balance=mortgage_payoff_native,
                is_primary_residence=False,  # toggled in UI by user editing tax_overrides if PPOR
                settings=settings,
            )

            gross_p = convert(tx["sale_price"], p.currency, primary_ccy, settings)
            tax_p = convert(tx["tax"], p.currency, primary_ccy, settings)
            net_p = convert(tx["net_to_owner"], p.currency, primary_ccy, settings)

            label = f"{p.name} — sell {fraction*100:.0f}% @ ~{sale_price_native:,.0f} {p.currency}"
            detail = dict(tx)
            detail["fraction_sold"] = fraction
            detail["holding_period_years"] = round(holding_period, 2)

            lines.append(
                FundingLineResult(
                    kind="property",
                    asset_label=label,
                    gross_proceeds=gross_p,
                    tax=tax_p,
                    net_proceeds=net_p,
                    detail=detail,
                )
            )
            continue

    total_gross = sum(l.gross_proceeds for l in lines)
    total_tax = sum(l.tax for l in lines)
    total_net = sum(l.net_proceeds for l in lines)

    return ProjectEvaluation(
        project_id=project.id,
        project_name=project.name,
        scenario_id=scenario.id,
        scenario_name=scenario.name,
        target_date=project.target_date,
        primary_currency=primary_ccy,
        total_cost=total_cost,
        funding_lines=lines,
        total_gross=total_gross,
        total_tax=total_tax,
        total_net_funding=total_net,
        surplus_or_shortfall=total_net - total_cost,
    )

File: lib/projects-engine.ts
/**
 * Investment-project capital-adequacy + tax engine.
 *
 * The user lists assets they'd be willing to use as funding (in priority
 * order). The engine walks the list, working out the smallest amount of
 * each asset needed to cover the project's net-of-tax cost. If the list
 * is exhausted with cost remaining, that's the shortfall.
 */

import { convert } from "./fx";
import {
  parseISO,
  vestedSharesAt,
  type InvestmentProject,
  type Property,
  type Scenario,
  type Settings,
  type StockHolding,
} from "./models";
import { projectPropertyValueAt, projectStockValueAt } from "./projections";
import { propertySaleTax, stockSaleTax, yearsBetween } from "./tax";

export type FundingLineResult = {
  kind: "stock" | "property" | "cash";
  asset_label: string;
  gross_proceeds: number;
  tax: number;
  net_proceeds: number;
  detail: Record<string, unknown>;
};

export type ProjectEvaluation = {
  project_id: string;
  project_name: string;
  scenario_id: string;
  scenario_name: string;
  target_date: string | null | undefined;
  primary_currency: string;
  total_cost: number;
  funding_lines: FundingLineResult[];
  total_gross: number;
  total_tax: number;
  /** Net proceeds actually drawn down to cover the project (≈ total_cost when funded). */
  total_net_funding: number;
  /** Full post-tax capacity of every listed source whether drawn or not. */
  total_available_net: number;
  /** total_available_net − total_cost. Positive = surplus capacity left over after the project. */
  surplus_or_shortfall: number;
  is_funded: boolean;
};

function resolveTargetDate(p: InvestmentProject): Date {
  if (p.target_date) {
    const d = parseISO(p.target_date);
    if (d) return d;
  }
  return new Date();
}

const fmt0 = new Intl.NumberFormat(undefined, { maximumFractionDigits: 0 });
const fmt2 = new Intl.NumberFormat(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 });

export function evaluateProject(args: {
  project: InvestmentProject;
  scenario: Scenario;
  holdings: StockHolding[];
  properties: Property[];
  settings: Settings;
}): ProjectEvaluation {
  const { project, scenario, holdings, properties, settings } = args;
  const target = resolveTargetDate(project);
  const primary = settings.primary_currency;

  // Total cost in primary currency
  let totalCost = 0;
  for (const item of project.items) {
    totalCost += convert(item.cost, item.currency, primary, settings);
  }

  const holdingsById = new Map(holdings.map((h) => [h.id, h]));
  const propsById = new Map(properties.map((p) => [p.id, p]));

  const lines: FundingLineResult[] = [];
  // How much net-of-tax funding we still need. Counts down as each source
  // contributes; once it hits 0, later sources emit zero-use lines.
  let remaining = totalCost;
  // Sum of every listed source's full post-tax capacity, whether drawn or
  // not — drives the headline surplus figure.
  let totalAvailableNet = 0;

  for (const fs of project.funding) {
    if (fs.kind === "cash") {
      // Cash is the only kind where the user must say *how much* they
      // have — it's not derivable from any tracked asset.
      const cashAvail = Math.max(0, fs.amount_or_shares);
      const use = remaining > 0 ? Math.min(remaining, cashAvail) : 0;
      totalAvailableNet += cashAvail;
      lines.push({
        kind: "cash",
        asset_label: `Cash · ${fmt0.format(cashAvail)} ${primary} available`,
        gross_proceeds: use,
        tax: 0,
        net_proceeds: use,
        detail: {
          cash_available: cashAvail,
          cash_used: use,
          cash_remaining: cashAvail - use,
          fully_used: cashAvail > 0 && use >= cashAvail,
          note: use === 0 && remaining <= 0 ? "Not needed — project already funded." : undefined,
        },
      });
      remaining -= use;
      continue;
    }

    if (fs.kind === "stock") {
      const h = holdingsById.get(fs.asset_id || "");
      if (!h) {
        lines.push({
          kind: "stock",
          asset_label: "(missing)",
          gross_proceeds: 0,
          tax: 0,
          net_proceeds: 0,
          detail: { error: `Holding ${fs.asset_id} not found` },
        });
        continue;
      }

      const v = projectStockValueAt(h, scenario, target, primary, settings);
      const projectedNative = v.projected_price;
      const liquidAvail = vestedSharesAt(h, target);
      const costBasis = h.cost_basis_per_share > 0 ? h.cost_basis_per_share : h.current_share_price;
      const holdingPeriod = Math.max(
        0,
        (target.getTime() - Date.now()) / (1000 * 60 * 60 * 24 * 365.25),
      );

      if (liquidAvail <= 0 || projectedNative <= 0) {
        lines.push({
          kind: "stock",
          asset_label: `${h.ticker || h.company_name}`,
          gross_proceeds: 0,
          tax: 0,
          net_proceeds: 0,
          detail: {
            note: liquidAvail <= 0 ? "No liquid shares at target date." : "Projected price ≤ 0.",
            shares_available: liquidAvail,
            projected_price_native: projectedNative,
          },
        });
        continue;
      }

      // For RSUs, apply the scenario's RSU income-tax dropdown as the all-in
      // haircut (matches the projections chart's net-of-tax line). Non-RSU
      // equity goes through the full CGT engine.
      const isRsu = h.equity_type === "RSU";
      const rsuRate = Math.max(0, Math.min(100, scenario.rsu_tax_rate_pct ?? 0)) / 100;

      // Run the tax calc once for the full liquid pool — proceeds and tax
      // are both linear in shares, so net-per-share derived from the full
      // pool is exact for any sub-quantity.
      let txFull = stockSaleTax({
        jurisdiction: project.jurisdiction,
        sale_price_per_share: projectedNative,
        cost_basis_per_share: costBasis,
        shares: liquidAvail,
        holding_period_years: holdingPeriod,
        settings,
      });
      if (isRsu) {
        const grossFull = projectedNative * liquidAvail;
        const taxFull = grossFull * rsuRate;
        txFull = { ...txFull, gross_proceeds: grossFull, tax: taxFull, net_proceeds: grossFull - taxFull };
      }
      const netFullPrimary = convert(txFull.net_proceeds, h.currency, primary, settings);
      totalAvailableNet += netFullPrimary;
      const netPerShare = netFullPrimary / liquidAvail;

      let sharesNeeded = 0;
      if (remaining > 0 && netPerShare > 0) {
        sharesNeeded = Math.min(liquidAvail, Math.ceil(remaining / netPerShare));
      }

      let tx = stockSaleTax({
        jurisdiction: project.jurisdiction,
        sale_price_per_share: projectedNative,
        cost_basis_per_share: costBasis,
        shares: sharesNeeded,
        holding_period_years: holdingPeriod,
        settings,
      });
      if (isRsu) {
        const grossN = projectedNative * sharesNeeded;
        const taxN = grossN * rsuRate;
        tx = { ...tx, gross_proceeds: grossN, tax: taxN, net_proceeds: grossN - taxN };
      }

      const gross_p = convert(tx.gross_proceeds, h.currency, primary, settings);
      const tax_p = convert(tx.tax, h.currency, primary, settings);
      const net_p = convert(tx.net_proceeds, h.currency, primary, settings);

      const label = sharesNeeded > 0
        ? `${h.ticker || h.company_name} — ${fmt0.format(sharesNeeded)} sh @ ~${fmt2.format(projectedNative)} ${h.currency}`
        : `${h.ticker || h.company_name} — not needed`;

      const detail: Record<string, unknown> = { ...tx };
      detail.projected_price_native = projectedNative;
      detail.shares_available = liquidAvail;
      detail.shares_used = sharesNeeded;
      detail.shares_remaining = liquidAvail - sharesNeeded;
      detail.fully_used = sharesNeeded >= liquidAvail;
      if (isRsu) {
        detail.tax_model = `RSU income tax @ ${(rsuRate * 100).toFixed(0)}% (scenario)`;
      }
      if (remaining > 0 && netPerShare > 0 && sharesNeeded === liquidAvail) {
        detail.warning = `Required all ${fmt0.format(liquidAvail)} liquid shares — still need more funding.`;
      }
      if (remaining <= 0) detail.note = "Not needed — project already funded.";

      lines.push({
        kind: "stock",
        asset_label: label,
        gross_proceeds: gross_p,
        tax: tax_p,
        net_proceeds: net_p,
        detail,
      });
      remaining -= net_p;
      continue;
    }

    if (fs.kind === "property") {
      const p = propsById.get(fs.asset_id || "");
      if (!p) {
        lines.push({
          kind: "property",
          asset_label: "(missing)",
          gross_proceeds: 0,
          tax: 0,
          net_proceeds: 0,
          detail: { error: `Property ${fs.asset_id} not found` },
        });
        continue;
      }

      const vNative = projectPropertyValueAt(p, scenario, target, p.currency, settings);
      const holdingPeriod = yearsBetween(p.purchase_date, target);

      // Net-to-owner scales linearly with fraction sold (sale price, cost
      // basis, and mortgage payoff all scale; the rate is flat). Compute
      // for fully-sold then derive fraction needed.
      const txFull = propertySaleTax({
        jurisdiction: project.jurisdiction,
        sale_price: vNative.gross,
        cost_basis: p.purchase_price,
        holding_period_years: holdingPeriod,
        mortgage_balance: p.mortgage_balance,
        is_primary_residence: false,
        settings,
      });
      const netFullPrimary = convert(txFull.net_to_owner, p.currency, primary, settings);
      totalAvailableNet += Math.max(0, netFullPrimary);

      if (netFullPrimary <= 0) {
        lines.push({
          kind: "property",
          asset_label: `${p.name}`,
          gross_proceeds: 0,
          tax: 0,
          net_proceeds: 0,
          detail: {
            note: "No equity at target date (mortgage ≥ projected value, or sale yields nothing after tax).",
            projected_value_native: vNative.gross,
          },
        });
        continue;
      }

      let fraction = 0;
      if (remaining > 0) {
        fraction = Math.min(1, remaining / netFullPrimary);
      }

      const salePriceNative = vNative.gross * fraction;
      const costBasisNative = p.purchase_price * fraction;
      const mortgageNative = p.mortgage_balance * fraction;

      const tx = propertySaleTax({
        jurisdiction: project.jurisdiction,
        sale_price: salePriceNative,
        cost_basis: costBasisNative,
        holding_period_years: holdingPeriod,
        mortgage_balance: mortgageNative,
        is_primary_residence: false,
        settings,
      });

      const gross_p = convert(tx.sale_price, p.currency, primary, settings);
      const tax_p = convert(tx.tax, p.currency, primary, settings);
      const net_p = convert(tx.net_to_owner, p.currency, primary, settings);

      const pctSold = fraction * 100;
      const label = fraction > 0
        ? `${p.name} — sell ${pctSold.toFixed(1)}% @ ~${fmt0.format(salePriceNative)} ${p.currency}`
        : `${p.name} — not needed`;

      const detail: Record<string, unknown> = {
        ...tx,
        fraction_used: fraction,
        fraction_remaining: 1 - fraction,
        fully_used: fraction >= 1,
        holding_period_years: Number(holdingPeriod.toFixed(2)),
      };
      if (remaining > 0 && fraction === 1) {
        detail.warning = `Fully selling property only yields ${fmt0.format(netFullPrimary)} ${primary} — still need more funding.`;
      }
      if (remaining <= 0) detail.note = "Not needed — project already funded.";

      lines.push({
        kind: "property",
        asset_label: label,
        gross_proceeds: gross_p,
        tax: tax_p,
        net_proceeds: net_p,
        detail,
      });
      remaining -= net_p;
      continue;
    }
  }

  const totalGross = lines.reduce((s, l) => s + l.gross_proceeds, 0);
  const totalTax = lines.reduce((s, l) => s + l.tax, 0);
  const totalNet = lines.reduce((s, l) => s + l.net_proceeds, 0);

  return {
    project_id: project.id,
    project_name: project.name,
    scenario_id: scenario.id,
    scenario_name: scenario.name,
    target_date: project.target_date,
    primary_currency: primary,
    total_cost: totalCost,
    funding_lines: lines,
    total_gross: totalGross,
    total_tax: totalTax,
    total_net_funding: totalNet,
    total_available_net: totalAvailableNet,
    // Surplus reflects the leftover capacity from EVERY listed source
    // (whether drawn down or not), so the user can see how much of their
    // available pool remains after the project.
    surplus_or_shortfall: totalAvailableNet - totalCost,
    // Allow a tiny rounding tolerance — sharesNeeded uses Math.ceil, so
    // we can be a fraction of a share over the cost.
    is_funded: totalAvailableNet >= totalCost - 0.01,
  };
}
